[1단계 - 행운의 로또 미션] 우디(박연기) 미션 제출합니다.

yeongipark opened the pull request:
# 인사
안녕하세요! 하리🐣 저는 우디(박연기)라고 합니다:)
먼저... pr작성에 앞서 **죄송하다는🙏 말씀**을 드리고 싶습니다...😭😭😭😭😭
저희가 다양한 에러를 효율적으로 처리할 방법을 고민하다보니, 로직이 많이 복잡해지고 읽기 어려운 코드가 되었습니다.
Validator가 Error 메시지를 결정하는게 맞나? 라는 의문점이 들었습니다. 여기서 시작한 리팩토링 점점 거대한 산이 되었습니다....
나름의 확장성과 유지보수성을 생각하며 설계를 하였으나, 쓸데없이 복잡한 부분이 큰 것 같습니다. 그래서 아래에 설계를 적어둘 예정이니 코드를 보기 전 한 번 참고해주셨으면 좋겠습니다. ㅠㅠ 😭😭😭

## 학습 목표
이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.
- UI와 도메인 영역을 분리할 수 있는 설계를 고민해보고, 목적에 맞게 객체와 함수를 활용
- TDD 방식으로 개발하며, 단위 테스트 기반으로 점진적인 리팩터링

## 제출 전 체크 리스트  
<!-- 리뷰를 요청하기 전, 다음 항목들을 기본적으로 확인해 주세요. -->
- [X] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [X] 기본적인 프로그래밍 요구 사항을 준수하고 있는지 확인했나요?
- [X] 테스트 코드는 모두 정상적으로 실행되나요?
- [ ] (해당하는 경우) 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포한 링크 기입: 

## 리뷰 요청 & 논의하고 싶은 내용  
<!-- PR 작성자로서, 코드만으로는 알기 어려운 작성자의 의도와 문제 해결 과정에 대해 공유해 주세요.  
이 PR에서 달성해야 하는 학습 목표를 잘 달성하고 있는지 스스로 확인하며, 피드백 받을 수 있는 내용일지 점검해 보세요.
리뷰어에게 정답을 묻기보다 고민하고 의사 결정한 과정에 대해 공유하고 이에 대한 피드백을 받으며 대화해보기를 권장합니다. -->

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
<!-- 구현 과정에서 가장 어려웠던 점이나 많이 고민한 점은 무엇인가요?
이를 해결하기 위해 어떤 방법들을 검토하고 시도했으며, 그 과정에서 새롭게 배운 점이 있나요? --> 
- 


### 2) 이번 리뷰를 통해 논의하고 싶은 부분  
<!-- 구현한 코드와 학습 목표와 관련해 피드백을 받고 싶은 부분이나, 함께 논의해보고 싶은 점 -->
- LottoMachine 이대로 괜찮은가?
   - 저희가 생각한 LottoMachine은 Lotto, Winnings 클래스와 Main(Controller) 클래스 사이에서 중간 계층 역할을 수행하도록 설계되었습니다. 그러나 구현을 진행하면서 Main 클래스가 LottoMachine의 역할 일부를 대신하게 되어, LottoMachine의 역할이 모호해진 느낌입니다.
하리는 LottoMachine의 역할을 명확하게 분리할 필요가 있는지, 그리고 어떤 부분을 신경 써서 리팩토링하면 좋을지에 대해 궁금합니다.
- TDD가 제대로 되었는가?
  - TDD를 적용하기 위해, 의존성이 가장 낮은 Lotto 클래스부터 단위 테스트를 작성하며 진행했습니다. TDD를 처음 접하는 입장에서 이러한 접근 방식이 맞는지 계속 고민하며 구현을 이어나갔습니다. 단위 테스트가 가능한 모든 로직에 대해 TDD를 적용하려고 했는데요, 혹시 TDD 사용 방식에서 잘못된 부분이나 누락된 부분이 있다면 조언 부탁드립니다!
- Validator의 역할
  - Validator는 입력 오류 여부만 판단하고, 구체적인 에러 메시지는 View에서 결정하는 것이 좋다는 피드백을 받았습니다. 그러나 저는 Validator가 검증 역할 전반을 수행해야 한다고 생각하여, 오류 발생 시 throw Error로 에러 메시지를 함께 전달해도 괜찮다고 보고 있습니다. 하리는 이에 대해 어떻게 생각하시는지 궁금합니다!

### 3) 로직 설명
우선, 가장 난해하게 작성된 부분이 InputHandler와 Validator라 생각해서 두 부분에 대해서 설명을 드리려고합니다...ㅠㅠ
#### a. InputHandler와 Validator의 존재 이유
> '꼭 Validator와 input 함수가 에러로 소통해야 할까?'라는 생각이 들었습니다!
잘못된 입력을 알려주는 것도 View 영역이라면, 판단은 Validator가 하고, 어떤 문자열을 보내줄지는 input 함수에서 책임지는 방법은 어떤가요?
이게 제가 지난 미션에서 받았던 코드 리뷰입니다. 이 부분에 대해서 확실히, 검증은 검증만 하고 view는 입출력만 하면되지 않나? 그렇다면, 에러 메세지에 따른 결정은 view가 해야하지 않을까? 하는 생각이 들었습니다.
그리고, 이번 미션에서 추가된 요구사항이 throw Error를 발생시켜야한다는 것이었는데 이 에러를 발생시키는 일을 Validator나 View가 하지 말고 유틸함수를 호출해 컨트롤러에서 발생해야하지 않을까 생각했습니다.
그래서 throwError라는 유틸함수가 추가로 만들어졌습니다.
#### b. InputHandler와 Validator의 로직 설명
먼저, 인풋핸들러와 벨리데이터는 객체 에러 코드로 소통합니다. 검증하는 로직은 계속 반복되기 때문에 각 검증 로직 키에 boolean 값을 받아, 어떤 키에 true값이 켜졌는지에 따라서 해당 인풋에 따라 미리 정의되어있는 해당 인풋메서드 전용 상수 메세지에 접근해 해당 에러메세지를 결정하는 식입니다. 이를 위해 키값의 조회가 잦게 일어났습니다.
- ***Validator**
먼저 Validator는 불리언을 반환하는 검사 로직의 모음집인 validationUtils 라는 객체가 정의되어있습니다.
그리고, Validator는 이 유틸에서 로직을 재활용하며 각 인풋필드에 맞는 메서드를 갖고 있습니다.
이렇게 한 이유는 입력필드마다 다른 유효성 검사를 진행해야하기 때문에, view나 controller에서 검증을 진행할 때 그에 맞는 메서드 하나만 호출하는 것이 좋을 것 같았기 때문입니다.
```js
 bonusNumber: (number) => {
    const errorResults = {
      IS_NUMBER_RANGE_OVER: ValidationUtils.isNumberRangeOver(number, 1, 45),
      IS_NOT_NATURAL_NUMBER: ValidationUtils.isNotNaturalNumber(number),
    };
    return errorResults;
  },
```
Validator 안의 메서드는 이렇게 errorResult를 반환하는데, 이는 에러 코드를 객체로 반환하는 방식입니다. 해당 값에 true가 하나라도 켜지면 에러가 있는 것으로 간주하고 에러를 발생시키는 식으로 에러 핸들링을 하였고,
저 에러리절트의 키와 동일한 키를 갖는 메세지 객체를 상수파일에 저장해, 해당하는 메세지를 출력하도록 하였습니다.
아래는 ERROR 메세지들을 저장한 객체입니다.
```js
const ERROR = {
  USER_INPUT: {
    IS_EMPTY: "[ERROR] 빈 값은 입력할 수 없습니다.",
  },
  WINNING_NUMBERS: {
    IS_WRONG_ARRAY_LENGTH: "[ERROR] 로또는 6개의 숫자로 이루어져야합니다.",
    IS_DUPLICATED_NUMBER: "[ERROR] 중복된 숫자는 입력하실 수 없습니다.",
    IS_ARRAY_NUMBER_RANGE_OVER: "[ERROR] 1~45 사이의 숫자를 입력해야합니다.",
    IS_NOT_NATURAL_NUMBER_IN_ARRAY: "[ERROR] 숫자는 자연수여야 합니다.",
  },
  BONUS_NUMBER: {
    IS_NUMBER_RANGE_OVER: "[ERROR] 1~45 사이의 숫자를 입력해야합니다.",
...
export default ERROR;
```
그리고, Output이 Validator에서 에러 객체를 받아, ERROR 메세지에 접근해 출력 처리하는 로직 부분입니다.
```js
  printErrorResults(errorResults, errorName) {
    Object.entries(errorResults).forEach(([key, value]) => {
      if (value) this.print(`${ERROR[errorName][key]}`);
    });
  },
```
- ***InputHandler**
inputHandler는 try...catch를 이용한 재귀함수 방식과, 사용자의 input 값을 빈값 검증->파싱->유효성 검증->return 하는 방식이 모든 input마다 반복되어 템플릿화하기 위해 만들어진 함수입니다.
```js
export async function inputHandler({
  inputMethod,
  parser,
  validatorMethod,
  errorName,
}) {
  try {
    const userInput = await userInputEmptyHandler(inputMethod);
    const parsedUserInput = parser ? Parser[parser](userInput) : userInput;
    const parsedUserInputError = Validator[validatorMethod](parsedUserInput);
    Output.printErrorResults(parsedUserInputError, errorName);
    throwError(parsedUserInputError);
    return parsedUserInput;
  } catch (error) {
    return inputHandler({
      inputMethod,
      parser,
      validatorMethod,
      errorName,
    });
  }
}
```
인풋핸들러의 로직은 이러합니다.
인자로 실행할 input함수와, 해당 인풋의 파서를 받고 어떤 유효성 검증을 해야하는지 validatorMethod 명을 받습니다. 그리고 ERROR 메세지를 출력할 때, 해당 인풋을 위한 에러메세지를 모아둔 상수에 키로 접근하기 위해 errorName을 받습니다.
그러면 순차적으로 빈값 검증->파싱->유효성 검증->return 이 실행됩니다.
이렇게 하면 장점이
1. 유효성 검증이 어긋난 모든 에러의 메세지를 출력할 수 있습니다.
2. 같은 검증 오류(보너스 넘버 입력과 당첨번호 입력을 할 때 둘다 1~45 사이의 숫자인지를 검증하는 등)를 반환해도 다른 메세지 처리를 할 수 있습니다.
3. 반복되는 input 로직을 템플릿화 할 수 있었습니다.
이러한 장점을 노리고 작성하였으나... 쓸데없이 복잡한 코드가 완성된 것 같습니다 ㅠㅠ 이 부분은 좀 더 리팩토링 진행해볼 생각입니다.

---

<details>
<summary>✅ 리뷰어 체크 포인트</summary>
<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.-->

### 1단계
- [X] TDD를 활용해 기능을 구현하는 과정에서 적절한 테스트 우선 접근 방식을 적용했는가? 단위 테스트의 커버리지는 충분한가?
- [x] 도메인과 UI의 관심사를 분리하여 적절한 모듈화가 이루어졌는가? 하나의 객체나 모듈이 너무 많은 책임을 가지고 있지는 않은가?
- [X] 객체의 프로퍼티를 직접 조작하기보다 메시지를 던지고 있는가?
- [x] 불필요한 클래스를 사용하지 않고, 함수를 적극적으로 활용하여 JavaScript다운 방식으로 로직을 구현했는가?

### 2단계 
- [ ] 도메인 로직에 불필요한 영향을 주지 않고 UI 변경에 대응했는가?
- [ ] DOM 조작과 이벤트 활용을 JavaScript의 개념에 맞게 이해하고 적절하게 적용했는가? 
- [ ] 웹 표준을 준수하는 마크업을 활용하며, 스타일 작성에 일관성이 있는가?
</details>

dev-dino22 committed:
docs(README.md): 기능목록 작성

dev-dino22 committed:
feat: 로또 랜덤 생성기 구현

dev-dino22 committed:
feat: 로또 객체 생성(generateLottos() 로또객체 내부로 이동)

dev-dino22 committed:
feat(Winnings): 당첨 번호와 보너스 번호를 검증한 후 저장

dev-dino22 committed:
feat(Winnings): 당첨 통계 계산

dev-dino22 committed:
feat(LottoMachine): 로또머신에 배열 생성 메서드 구현

dev-dino22 committed:
refactor(LottoMachine): getStatistics() 메서드 삭제

dev-dino22 committed:
feat(Validator):메서드  유효성 검사 로직 작성

dev-dino22 committed:
refactor(Winnings): 검증 로직 리팩토링

dev-dino22 committed:
feat: View 작성

dev-dino22 committed:
feat(Error): 에러 메세지 객체 생성

dev-dino22 committed:
refactor: 메세지 상수화

dev-dino22 committed:
feat(Main): 메인 로직 작성

dev-dino22 committed:
refactor: Input, Output 수정

dev-dino22 committed:
feat: Index 작성

dev-dino22 committed:
test: console.log 제거

LAH1203 commented on README.md:
제가 다른 분들께도 강조 드리는 부분인데요! 개발자는 개발을 잘 하는 것도 중요하지만 문서화를 잘 하는 것도 그만큼 중요하다고 생각합니다. 문서는 나보다는 남이 읽는다는 생각으로 정리하면 좋아요.

근데 우디가 적어주신 리드미를 보면 아래 이미지처럼 보여져요.

<img width="298" alt="image" src="https://github.com/user-attachments/assets/80f5818f-c8f8-4f92-8b37-48a9ca5ddf6b" />

지금 적어주신 것은 본인이 보기 위한 단순 나열로밖에는 안 보이는 것 같아요 ㅠㅠ 저는 문서만 보고도 어떤 방식으로 설계했는지 알 수 있도록 만들어야 한다고 생각합니다~

어떻게 적으면 우디의 의도를 잘 전달할 수 있을지 한번 정리해보면 어떨까요? 지금보다 충분히 잘 정리해주실 수 있을 것 같아요~!

yeongipark replied:
네!!

LAH1203 commented on src/view/Output.js:
```diff
@@ -0,0 +1,45 @@
+import { OUTPUT_MESSAGE } from "../constant/Message.js";
+
+const Output = {
+  print(message) {
+    console.log(message);
+  },
```
print 함수는 단순히 console.log의 래핑으로밖에 안 보이는 것 같아요. 이렇게 따로 만드신 이유가 있을까요?

yeongipark replied:
console.log()를 그대로 사용하는 것보다, print 메서드를 사용하는게 직관적으로 이해할 수 있다고 생각하여 만들었습니다!

하지만 지금 다시 생각해보니, console.log()메서드는 너무 유명하고 모두 알고 있는 메서드라서 그대로 사용해도 된다고 생각이 드네요!

LAH1203 replied:
저는 다양한 시도를 해보시는 건 좋다고 생각해요~! 다른 시각에서 한번 생각해봐도 좋을 것 같아서 코멘트를 남겼었어요 :)

LAH1203 commented on src/view/Output.js:
```diff
@@ -0,0 +1,45 @@
+    console.log(message);
+  },
+
+  lottoAmount(lottoAmount) {
+    this.print(OUTPUT_MESSAGE.LOTTO_AMOUNT(lottoAmount));
+  },
+
+  lottoNumbers(lottoNumbers) {
+    const copyLottoNumbers = [...lottoNumbers];
+    copyLottoNumbers.sort((a, b) => a - b);
+    this.print(`[${copyLottoNumbers.join(DEFINITION.SPLIT)}]`);
+  },
+
+  winningStatistics() {
+    this.print(OUTPUT_MESSAGE.WINNING_STATISTICS);
+    this.print(OUTPUT_MESSAGE.BOUNDARY);
+  },
+
+  newLine() {
+    this.print(DEFINITION.EMPTY);
+  },
+
+  matchResult(rank, amount) {
+    const prize = DEFINITION.LOTTO_PRISE[rank].toLocaleString();
+    this.print(OUTPUT_MESSAGE.MATCH_RESULT(rank, amount, prize));
+  },
+
+  winningRate(rate) {
+    this.print(OUTPUT_MESSAGE.WINNING_RATE(rate));
+  },
+
+  printErrorResults(errorResults, errorName) {
+    Object.entries(errorResults).forEach(([key, value]) => {
+      if (value) this.print(`${ERROR[errorName][key]}`);
+    });
+  },
```
함수 네이밍 간의 규칙이 정해지면 좋을 것 같아요! 지금은 별도의 규칙이 없이 명사형, 동사형이 모두 섞여 있는 형태네요

yeongipark replied:
네!

LAH1203 commented on src/view/Output.js:
```diff
@@ -0,0 +1,45 @@
+  },
+
+  lottoNumbers(lottoNumbers) {
+    const copyLottoNumbers = [...lottoNumbers];
+    copyLottoNumbers.sort((a, b) => a - b);
+    this.print(`[${copyLottoNumbers.join(DEFINITION.SPLIT)}]`);
```
정렬된 새로운 배열이 필요하다면 `toSorted`를 활용해볼 수 있을 것 같아요~

근데 애초에 가지고 있을 때 정렬해서 가지고 있는 게 아니라 출력할 때 정렬해서 출력하는 방식이네요...! 이렇게 만드신 이유가 궁금합니당

yeongipark replied:
toSorted라는 메서드가 있었군요..! 감사합니다:)

Lotto 클래스에서 로또 번호를 저장할 때 오름차순으로 정렬할 수 있었지만, 혹시 랜덤으로 만들어진 배열 자체가 필요한 경우가 있을 수도 있다 생각했고, 또 오름차순으로 정렬하여 출력이 view의 역할이라 생각하여 view에서 오름차순으로 정렬하여 출력했습니다..!

하리는 Lotto클래스에 로또 번호를 저장할 때, 오름차순으로 정렬하여 저장해도 괜찮다고 생각하시나요?

LAH1203 replied:
저는 오름차순으로 정렬해서 저장해도 상관은 없을 것 같아요~! 근데 지금 우디가 적용해주신 방식이 틀려서 고쳐야 한다는 건 절대 아니에요! ㅋㅋㅋㅋ 저는 다양한 시각에서 코드를 바라보면 좋을 것 같아서 남겼었어요

이 부분을 프론트 - 백의 역할로 나눠서 생각해봐도 좋을 것 같아요. 저희가 API를 통해 데이터를 받아와 화면에 그려줄 때를 가정해볼게요! 프론트에서 정렬을 하는 경우가 많을까요, 아님 백에서 애초에 정렬해서 데이터를 보내주는 경우가 많을까요?! 이런 관점에서 정렬을 어디에서 수행하는 게 맞을지 생각해봐도 좋아요~

yeongipark replied:
오..! 

프로젝트를 진행했을 때, 백에서 정렬해서 받은 적도 있고 프론트에서 정렬해서 보여준 적도 있었어요!! 

프론트를하는 저의 입장에서는 백에서 정렬해서 주는게 훨씬 편했지만 이는 정말 정답이 없고 팀에 따라 달라질 수 있는 부분이네요!! 
감사합니다:)

LAH1203 commented on src/view/Input.js:
```diff
@@ -0,0 +1,51 @@
+import readline from "readline";
+import { INPUT_MESSAGE } from "../constant/Message.js";
+
+const Input = {
```
전체적으로 함수를 객체로 묶어서 사용하신 경우가 많은 것 같더라고요.
그렇게 만드신 이유와 장점이 궁금해요~

yeongipark replied:
객체로 묶어 사용하면 개별 함수를 export하는 것보다 import 구문을 줄일 수 있고, 비슷한 역할을 하는 메서드를 하나의 객체로 모아 응집도를 높일 수 있다고 생각했습니다.

예를 들어, inputUserInput이라는 함수가 필요할 경우 이를 독립된 함수로 export하여 사용할 수도 있지만, 이를 Input 객체의 메서드로 작성하면 Input.userInput과 같이 보다 직관적으로 사용할 수 있습니다. 또한, 해당 메서드가 어디서 정의 되었는지 쉽게 파악할 수 있다는 장점이 있다고 생각하여 객체로 묶어 사용했습니다!

LAH1203 commented on src/util/util.js:
```diff
@@ -0,0 +1,3 @@
+export function throwError(errorResult) {
+  if (Object.values(errorResult).some((value) => value)) throw new Error();
+}
```
이 throwError가 적용된 곳을 봤는데, try-catch 문으로 감싸서 에러 발생 시 재귀 함수를 돌리기 위함이네요. 근데 또 보면 출력은 에러 발생과 별개로 throwError 호출 전에 진행하고 있어요.
저는 이 구조 자체가 굉장히 낯설었어요. 일단 가장 큰 이유는 '왜 에러 발생과 에러 출력은 따로 실행되지?' 였습니다.
보통 에러가 발생하면 출력하는 플로우를 예측합니다. 근데 지금은 에러가 있는지 여부를 printErrorResults, throwError에서 각각 판단하고 있어요. 이건 함수가 잘못 분리되어 결합된 형태라고 생각합니다. 프로젝트 전체적으로 함수가 errorResult 객체에 너무 강하게 의존하고 있어요.

그래서 리팩토링을 진행하신다면 이 부분에 집중해서 어떻게 했을 때 errorResult 객체에 의존하지 않는 함수로 분리해 사용할 수 있을지, 지금의 함수 분리는 적절한지를 먼저 고민해보시면 좋을 것 같습니다.
덧붙여 PR에서 이렇게 구현했을 때의 장점으로 '유효성 검증이 어긋난 모든 에러의 메세지를 출력할 수 있습니다.' 라고 하셨는데, 사용자의 입장에서도 '입력에 대한 모든 에러 메시지를 한번에 확인할 필요가 있을까?' 생각해보셔도 좋고요 :)

yeongipark replied:
저도 작성하면서 이게 맞나...? 라는 생각을 했었던 로직입니다..!

이번 리팩토링때 하리의 리뷰 토대로 신경써서 수정해 보겠습니다!

감사합니다:)

LAH1203 commented on src/util/Parser.js:
```diff
@@ -0,0 +1,8 @@
+  toNumber: (string) => Number(string),
+  toSplitNumberArray: (stringNumbers) =>
+    stringNumbers.split(",").map((string) => Number(string)),
+  toYNBoolean: (string) => string.toLowerCase() === "y",
```
y나 n도 상수화를 진행할 수 있을 것 같아요~

yeongipark replied:
상수화 부분을 핑계지만.... 시간이 부족하여 많이 진행하지 못했씁니다..!

리팩토링 과정에서 완벽한 상수화를 하도록 하겠습니다!

LAH1203 commented on src/constant/Error.js:
```diff
@@ -0,0 +1,29 @@
+const ERROR = {
+  USER_INPUT: {
+    IS_EMPTY: "[ERROR] 빈 값은 입력할 수 없습니다.",
+  },
+  WINNING_NUMBERS: {
+    IS_WRONG_ARRAY_LENGTH: "[ERROR] 로또는 6개의 숫자로 이루어져야합니다.",
+    IS_DUPLICATED_NUMBER: "[ERROR] 중복된 숫자는 입력하실 수 없습니다.",
+    IS_ARRAY_NUMBER_RANGE_OVER: "[ERROR] 1~45 사이의 숫자를 입력해야합니다.",
+    IS_NOT_NATURAL_NUMBER_IN_ARRAY: "[ERROR] 숫자는 자연수여야 합니다.",
+  },
+  BONUS_NUMBER: {
+    IS_NUMBER_RANGE_OVER: "[ERROR] 1~45 사이의 숫자를 입력해야합니다.",
+    IS_NOT_NATURAL_NUMBER: "[ERROR] 숫자는 자연수여야 합니다.",
+    IS_DUPLICATED: "[ERROR] 당첨번호와 중복된 숫자는 입력하실 수 없습니다.",
+  },
+  WINNINGS_AND_BONUS: {
+    IS_DUPLICATED: "[ERROR] 당첨번호와 중복된 숫자는 입력하실 수 없습니다.",
+  },
+  PURCHASE_PRICE: {
+    IS_NUMBER_RANGE_OVER:
+      "[ERROR] 1000 이상 100억 이하의 숫자를 입력해야합니다.",
+    IS_NOT_MULTIPLE: "[ERROR] 1000 단위로 나누어 떨어지는 숫자여야합니다.",
+  },
+  RESTART: {
+    IS_NOT_YN: "[ERROR] 선택은 Y 또는 N 으로 입력해야합니다.",
+  },
+};
```
여기에서도 상수화할 수 있는 값들이 조금 보이는 것 같아요~

yeongipark replied:
네 감사합니다:)

LAH1203 commented on src/constant/definition.js:
```diff
@@ -0,0 +1,28 @@
+
+const DEFINITION = {
+  LOTTO_PRISE: {
+    [RANK.FIFTH]: 5000,
+    [RANK.FOURTH]: 50000,
+    [RANK.THIRD]: 1500000,
+    [RANK.SECOND]: 30000000,
+    [RANK.FIRST]: 2000000000,
```
큰 숫자에는 언더바를 활용해보면 어떨까요?

yeongipark replied:
언더바로 숫자를 구분할 수 있는 것을 모르고 있었네요!

수정하겠습니다!!

감사합니당

LAH1203 commented on src/domain/Validator.js:
```diff
@@ -0,0 +1,85 @@
+const ValidationUtils = {
+  isEmpty: (string) => string.trim().length === 0,
+  isCorrectedArrayLength: (array, min, max) =>
+    array.length < min || array.length > max,
+  isDuplicatedNumber: (array) => new Set(array).size !== array.length,
+  isArrayNumberRangeOver: (numbers, min, max) =>
+    numbers.some((number) =>
+      ValidationUtils.isNumberRangeOver(number, min, max)
+    ),
+  isNotNaturalNumberInArray: (numbers) =>
+    numbers.some((number) => ValidationUtils.isNotNaturalNumber(number)),
+  isNumberRangeOver: (number, min, max) => number < min || number > max,
+  isNotNaturalNumber: (number) => number % 1 !== 0 || number < 1,
+  isDuplicated: (array, element) =>
+    new Set([...array, element]).size !== array.length + 1,
+  isNotMultiple: (number, unit) => number % unit !== 0,
+  isYN: (string) =>
+    string.toLowerCase() !== "y" && string.toLowerCase() !== "n",
+};
```
모든 입력에 대한 검증 로직이 합쳐져 있어 가독성이 떨어지는 것 같아요 ㅠ_ㅠ
좀 더 분리하거나 다른 구조를 고려해볼 수는 없을까요?
아예 태초로 돌아가서 ValidationUtils 구조화는 필요할지! 요것도 같이 고민해봅시당

yeongipark replied:
흑흑 동의하는 부분입니다!

조금더 고민하여 리팩토링 하겠습니다!

LAH1203 commented on __tests__/LottoMachine.test.js:
```diff
@@ -0,0 +1,17 @@
+import Lotto from "../src/domain/Lotto.js";
+import Winnings from "../src/domain/Winnings.js";
+
+describe("LottoMachine.test.js", () => {
+  test("로또 배열을 저장함", () => {
```
테스트 코드의 설명 스타일이 전체적으로 통일되면 좋을 것 같습니당

yeongipark replied:
네!!

LAH1203 commented on src/controller/Main.js:
```diff
@@ -0,0 +1,101 @@
+import Winnings from "../domain/Winnings.js";
+import { inputHandler } from "../util/InputHandler.js";
+
+export default class Main {
```
어떤 건 클래스, 어떤 건 함수를 묶어놓은 객체로 정의되어 있네요. 따로 분리해서 적용하신 기준이 있을까요?

yeongipark replied:
필드를 가지지 않거나, 그냥 함수의 집합인 경우 객체로 선언하고

각각의 함수가 연관(프라이빗 메서드를 모아서 하나의 메서드로 선언하는 경우)되어 있거나, 필드가 있는 경우는 클래스로 작성했습니다!

하지만 이부분에 대해서는 항상 고민입니다. 처음에는 정적 메서드만 가지는 클래스는 정적 클래스로 사용하습니다. 하지만 이를 객체로 바꿔 사용해도 괜찮다고 생각이 들어, 굳이 클래스를 사용해야 되나? 라는 생각에 정적 메서드만을 가지거나 필드를 가지지 않는 경우에는 객체로 선언했습니다!

이 부분에 대해서 하리의 생각이 궁금합니다!

LAH1203 replied:
> 처음에는 정적 메서드만 가지는 클래스는 정적 클래스로 사용하습니다. 하지만 이를 객체로 바꿔 사용해도 괜찮다고 생각이 들어, 굳이 클래스를 사용해야 되나? 라는 생각에 정적 메서드만을 가지거나 필드를 가지지 않는 경우에는 객체로 선언했습니다!

이 부분은 좋은 과정을 거치고 계신 것 같아요! 저도 비슷하게 생각하고 있습니다.

그렇다면 컨벤션 관점에서 생각해보셔도 좋아요! 보통 클래스는 파스칼 케이스로 네이밍하고, 함수는 카멜 케이스로 네이밍해요. 그럼 이 관점에서 봤을 때 연관된 함수를 모아놓은 객체를 네이밍할 때 파스칼 케이스와 카멜 케이스 중 어떤 게 좀 더 적절할까요? 정의한 객체를 사용하는 곳에서 그 객체를 사용하는 방식 (ex: `Validator.restart('n')`) 을 봤을 때, Validator는 상태를 가지지 않기에 클래스가 아닌 단순 객체로 선언된 형태인지를 알 수 있을까요?

yeongipark replied:
카멜 케이스로 작성하면 클래스로 생각할 수 있을거 같아요..!

하리의 질문을 토대로 제가 생각했을 때, 단순 함수의 집합의 객체의 경우에는 파스칼 케이스가 아닌 카멜 케이스로 작성하는 것이 형태를 파악하는데 유리하다고 생각합니다!!

LAH1203 commented on src/controller/Main.js:
```diff
@@ -0,0 +1,101 @@
+import Winnings from "../domain/Winnings.js";
+import { inputHandler } from "../util/InputHandler.js";
+
+export default class Main {
```
Main의 역할이 과중되어 있기는 하네요... 🥲

yeongipark replied:
main은 view와 도메인의 결합을 통해 프로그램 전체 흐름을 담당한다고 생각했습니다! 

하리가 보셨을 때, 어떤 부분에서 과중되어있다고 생각하시나요..?

LAH1203 replied:
`main은 view와 도메인의 결합을 통해 프로그램 전체 흐름을 담당한다고 생각했습니다!` 라고 하셨는데, 제가 봤을 때는 단순 결합 뿐만 아니라 도메인의 역할도 수행하고 있는 것으로 보여요~!

예를 들면, `inputBonus`에서 맞는 입력값이 들어올 때까지 검증을 반복하는 형태는 도메인의 역할이 아닐까요? `inputRestart`에서 y라는 입력값이 들어왔을 때 새 게임을 시작하는 것도 도메인의 역할이라고 볼 수 있지 않을까요?

이런 식으로 Main에서 정의된 메서드들을 보면, 다들 하나의 역할이 아닌 여러 개의 역할을 수행하고 있다고 느꼈습니다. 말씀해주신대로 Main이 "결합"의 역할을 한다면 지금보다 더 나눠지고 세분화되어야 한다고 생각해요. 그 관점에서는 결합 외의 로직은 Main 클래스 내에 존재할 이유가 없겠죠?!

LAH1203 commented on src/domain/Lotto.js:
```diff
@@ -0,0 +1,15 @@
+  }
+
+  static generateLottoNumbers() {
+    const randomNumbers = [];
+    while (randomNumbers.length < 6) {
+      const randomNumber = Math.floor(Math.random() * 44 + 1);
+      if (randomNumbers.includes(randomNumber)) continue;
+      randomNumbers.push(randomNumber);
+    }
+    return randomNumbers;
```
`Array.includes`는 매번 배열에 value가 있는지 체크해야 해서 여기에서는 이중 반복문을 도는 구조처럼 만들어질 것 같아요. 어떻게 개선해볼 수 있을까요?

yeongipark replied:
Set 자료구조를 사용해 볼 수 있을거 같아요!!

LAH1203 commented on src/domain/LottoMachine.js:
```diff
@@ -0,0 +1,9 @@
+
+const LottoMachine = {
+  createLottos(money) {
+    return Array.from({ length: money / 1000 }).map(() => new Lotto());
```
전체적으로 코드에 매직 넘버에 대해 상수화가 잘 적용되지 않은 것 같아요 😭
한번 관련해서 정리해보시면 어떨까요?

yeongipark replied:
리팩토링 과정에서 상수화 진행하도록 하겠습니다,,,,,!

LAH1203 commented on src/domain/Winnings.js:
```diff
@@ -0,0 +1,86 @@
+  }
+
+  #isBonusWinning(matchCount, lotto) {
+    if (
+      matchCount === DEFINITION.LOTTO_RULE[RANK.SECOND] &&
+      lotto.includes(this.bonusNumber)
+    )
+      return true;
+    return false;
```
요런 코드는 간단하게 조건문을 바로 리턴하는 방식을 사용해볼 수 있습니당

```suggestion
    return matchCount === DEFINITION.LOTTO_RULE[RANK.SECOND] && lotto.includes(this.bonusNumber);
```

yeongipark replied:
훨씬 간결하네요!! 감사합니다:)

LAH1203 reviewed (changes requested):
안녕하세요 우디! 하리입니당~ 이번 미션 잘 부탁드려요 🫡

먼저 우디가 PR에 남겨주신 질문부터 답변을 드려볼게요~!

> LottoMachine 이대로 괜찮은가?
> 저희가 생각한 LottoMachine은 Lotto, Winnings 클래스와 Main(Controller) 클래스 사이에서 중간 계층 역할을 수행하도록 설계되었습니다. 그러나 구현을 진행하면서 Main 클래스가 LottoMachine의 역할 일부를 대신하게 되어, LottoMachine의 역할이 모호해진 느낌입니다. 
> 하리는 LottoMachine의 역할을 명확하게 분리할 필요가 있는지, 그리고 어떤 부분을 신경 써서 리팩토링하면 좋을지에 대해 궁금합니다.

저는 모든 객체는 역할이 명확하게 분리될 필요가 있다고 생각합니다. 그런데 현재 구조를 보면 LottoMachine뿐만 아니라 domain 폴더 내의 여러 객체(Main 포함)의 역할이 분명하게 나누어지지 않은 것 같습니다. (Lotto 객체는 예외입니다.)

- Validator는 Winnings, Main, InputHandler에서 각각 유틸 함수처럼 호출되는데, 검증 흐름이 한곳에서 단일하게 관리되지 않고 분산되어 있습니다.
- Winnings도 단순한 유틸 함수처럼 호출될 뿐만 아니라, Main과 InputHandler에서 이미 검증한 데이터를 다시 검증하고 있습니다.
- LottoMachine은 단순히 로또 객체 목록을 생성만 하는 반면, Main은 입출력 관리뿐만 아니라 검증까지 담당하고 있습니다.

이렇게 객체 간 책임이 명확하게 분리되지 않으면, 플로우(입력 → 검증 → 출력)가 코드에 명확하게 반영되지 않아서 유지보수가 어려워집니다.

그래서 설계를 다시 고민해보면 좋을 것 같습니다.

- 요구사항을 기준으로 어떤 기능이 필요한지 나열해보고,
- 기능을 묶을 때 단순히 비슷한 기능이라 묶는 것이 아니라, 진짜 관련성이 있는지 고려하면서 구조를 잡으면
- 지금보다 더 명확한 역할을 가진 객체로 나뉠 것이라고 생각합니다.

이런 방식으로 다시 정리해보면, LottoMachine의 역할도 좀 더 분명해질 것 같아요!

> TDD가 제대로 되었는가?
> TDD를 적용하기 위해, 의존성이 가장 낮은 Lotto 클래스부터 단위 테스트를 작성하며 진행했습니다. TDD를 처음 접하는 입장에서 이러한 접근 방식이 맞는지 계속 고민하며 구현을 이어나갔습니다. 단위 테스트가 가능한 모든 로직에 대해 TDD를 적용하려고 했는데요, 혹시 TDD 사용 방식에서 잘못된 부분이나 누락된 부분이 있다면 조언 부탁드립니다!

작은 단위부터 TDD를 시도하신 건 좋은 접근 방식이라고 생각합니다 ㅎㅎ

다만 모든 로직에 대해 TDD를 적용할 필요는 없을 것 같아요. '이 테스트가 실제로 의미가 있는지?'를 먼저 생각해보시면 좋지 않을까 생각합니다.

TDD는 어떤 방식으로 적용하셨는지도 알려주시면 좋을 것 같아요. 예를 들면, 테스트 코드를 작성하실 때 `Red → Green → Refactor` 순서를 잘 지키셨나요?

> Validator의 역할
> Validator는 입력 오류 여부만 판단하고, 구체적인 에러 메시지는 View에서 결정하는 것이 좋다는 피드백을 받았습니다. 그러나 저는 Validator가 검증 역할 전반을 수행해야 한다고 생각하여, 오류 발생 시 throw Error로 에러 메시지를 함께 전달해도 괜찮다고 보고 있습니다. 하리는 이에 대해 어떻게 생각하시는지 궁금합니다!

좋은 고민이네요!
저는 둘 다 좋은 방식이라고 생각합니다. 에러 메시지를 어디에서 관리하느냐에 따라서 장단점이 확실하게 나뉜다고 생각해요.

다만 좀 더 선호하는 건 에러 메시지를 View에서 관리하는 방식인 것 같아요. 저라면 특정 에러 발생 시 어떤 에러 메시지를 반환하는지는 검증에 대한 역할보다는 검증한 로직에 대한 후속 조치의 영역이라고 볼 것 같습니당

---

따로 말씀해주셨던 것처럼 리팩토링 관련해서 고민을 많이 하시는 것 같아서, 리팩토링 관점에서 코멘트를 많이 남겨봤어요! 객체 분리 관련해서는 각각 코멘트를 달기 애매해서, 달지 않고 질문에 대한 답변으로 제 생각을 적어봤습니다~ 우디의 의견을 편하게 남겨주시면 좋을 것 같아요 :-)

yeongipark committed:
refactor: validator 리팩토링

- ValidatorUtil을 util로 이동
- 반복 되어 사용되는 로직만 util로 분리
- 각 관심사에 대해 validator를 분리

LAH1203 commented:
@yeongipark 수정하신 커밋은 나중에 각각의 코멘트에 어떻게 반영하셨는지 함께 커밋 번호를 남겨주시면 좋을 것 같습니다~! ㅎㅎ 우디 화이팅 💪💪

yeongipark committed:
refactor(validator) : 바뀐 validator 적용

yeongipark committed:
refactor: errorHandler 방식 변경

- 기존의 복잡한 형태의 에러 핸들러 방식을 간단한 방식으로 변경

yeongipark committed:
refactor: 객체 변수명 카멜 케이스로 변경

yeongipark committed:
refactor(main): 다시 시작 부분도 errorHandler 방식으로 변경

yeongipark committed:
refactor: 프로그램의 전체적인 구조 변경

- Main 클래스가 LottoMachine 클래스만을 통해 상호작용할 수 있도록 변경
- Lottos 클래스를 만들어 로또 배열을 저장
- getter를 사용하는 것이 아닌 메시지를 보내 상호작용하도록 구현

yeongipark committed:
refactor: output 메서드 로직 변경

- Main 클래스의 로직을 단순화 하기 위해 ouput 메서드 로직 변경

yeongipark committed:
docs: README.md 파일 설계 추가

- 기능을 세분화 함
- 설계 내용을 추가함

yeongipark committed:
chore: 필요없는 파일 삭제

yeongipark committed:
refactor : 프로그램 상수화

yeongipark committed:
refactor: 응집도를 위해 상수를 사용하는 폴더에 상수 이동

yeongipark committed:
refactor: 큰 숫자 "_" 추가하여 가독성 증가

yeongipark committed:
refactor: 메서드 명 명사형으로 통일

yeongipark committed:
refactor: 덜한 상수화 마저 진행

yeongipark committed:
test(Lotto): Lotto 클래스 테스트 코드 수정

yeongipark committed:
test: 없어진 클래스 테스트 코드 삭제

yeongipark committed:
test: BonusNumberValidator 클래스 테스트 코드 작성

yeongipark committed:
test: Calculator 클래스 테스트 코드 작성

yeongipark committed:
test: Lotto 클래스 테스트 코드 작성

yeongipark committed:
test: NumbersValidator 클래스 테스트 코드 작성

yeongipark committed:
test: PurchasePriceValidator 클래스 테스트 코드 작성

yeongipark committed:
docs: README.md 파일에 설계도 이미지 추가

yeongipark commented:
안녕하세요 하리!
하리의 리뷰를 토대로 코드 리팩토링을 진행했습니다!!

잘 부탁드립니당 😊

--- 
# 🤔질문
### 1. 커밋 
- 특정 기능을 구현하는 과정에서 여러 파일을 보다가, 갑자기 어떤 파일에 잘못된 코드가 있거나 리팩토링할 여지가 있는 코드를 발견할 때, 해당 파일만 수정해서 바로 커밋하는 편이 좋을까요? 아니면 수정할 부분은 나중에 미루고, 우선 현재 진행 중인 기능 구현을 마친 후에 별도로 수정하는 게 좋을까요??
저는 매번 보일 때마다 수정해서 커밋이 뒤죽박죽이 되고 커밋 수도 많아지는 문제가 있는데, 이에 대해 어떻게 접근하는 게 좋은지 궁금합니다!

### 2. 재시작이 도메인 로직일까?
> y라는 입력값이 들어왔을 때 새 게임을 시작하는 것도 도메인의 역할이라고 볼 수 있지 않을까요?
- 하리의 리뷰에서 재시작 부분이 도메인 로직이라고 하신 부분이 있었어요! 이 부분에 대해 저도 고민해 봤는데, 재시작 여부를 사용자에게 입력받고 처리하는 것은 도메인 로직에서 처리하기 보다는 애플리케이션 흐름을 담당하는 Main 클래스가 처리하는게 맞다고 생각하여, Main 클래스에 해당 로직을 작성했습니다!
이 부분에 대해 하리의 생각이 궁금합니다!

### 3. 클래스 내부 메서드 위치 
- 클래스에 메서드가 많아지면, 퍼블릭 메서드와 함께 여러 프라이빗 메서드도 등장하게 됩니다. 저는 가독성을 위해 퍼블릭 메서드에서 사용되는 프라이빗 메서드를 해당 퍼블릭 메서드 바로 아래에 배치하는 방식을 선호합니다. 하지만 전체 코드를 넓게 보았을 때, 퍼블릭과 프라이빗 메서드가 섞여 있으면 다소 난잡해 보일 수 있습니다. 이런 경우에는 프라이빗 메서드끼리 모으고, 퍼블릭 메서드끼리 모아서 배치하는 방법도 고려해볼 만하다고 생각합니다!  하리께서는 클래스 내부의 메서드 배치는 어떤 식으로 구성하시는지 궁금합니다!

---

# ✍️수정한 부분
### 1. Validator 리팩토링
- 관련 커밋 : 1802fb302441fb54e0f739edff5973e4c7443e29, bea0a068b7b20111dff3eb18e5709e55e863b212
- 기존에는 모든 유효성 검증 로직을 하나의 객체에 저장하여 사용했는데, 이 방식은 특정 입력에 대한 검증 로직을 찾는 데 시간이 오래 걸리는 문제가 있었습니다. 그래서 저는 기능을 더 세분화하여, 각 입력에 대응하는 별도의 Validator 클래스를 만들었습니다. 객체 대신 클래스로 작성한 이유는 외부에 노출하면 안 되는 프라이빗 메서드들이 포함되어 있기 때문입니다. 또한, 재사용되는 로직만을 validationUtil 파일로 분리하여, 전체적으로 가독성을 높였습니다.
- 현재 유효성 검사는 여러 곳에서 수행되고 있습니다. Main 클래스에서는 사용자 입력에 대한 유효성 검사를 진행하지만, 특정 클래스 내부에서도 한 번 더 검증하는 이유는 예상치 못한 잘못된 입력으로부터 클래스 생성 자체를 막기 위함입니다. 프론트엔드는 보안에 취약한 영역이기 때문에, 클래스 내부에 유효성 검증 로직을 추가하여 보안을 강화했습니다.

### 2. 메시지 상수화
- 관련 커밋 : 63fceb0a62505bf748062e014167ecf5e072421b, c80a2f9d06319a7a10fa88696aaeed53f41181bc, 657c7f0276273eafb233fadefc370780ed456338
- 상수를 하나의 폴더에 작성하는 것이 아니라, 한 곳에서만 사용되는 상수의 경우에는 해당 파일 내부에 선언하여 응집도를 높였습니다!

### 3. 테스트 코드 수정 및 추가
- 관련 커밋 : d79a6d78f36539ce98e6d35a8971e723976a543c, 462cf739b224d66a04fef909ec1b6c34310f0ebe, 1c27eb698b0bcbb9f49260ba3d89a5e77589ef4a, 77de335f2285a2bf43419552e6fefa3f280bf4f3, 0e6e400942d4185e54a9db5ca9167cf27e772809, 2c79cb9bff7a151ebf0f18b5e4c5ea0686c2af93, 15877375c0f6897ddb1574de35c9c588efa0fc6a
- 테스트 코드 설명 작성 스타일을 통일했습니다!

### 4. 큰 숫자에 "_" 추가하여 가독성 높이기
- 관련 커밋 : 3e0da895f47131ffee3ee1b52f8ca8acedfb6c86

### 5. README.md 파일에 설계 내용 추가
- 관련 커밋 : d83aee48cd16d406461db498e11eaed528111ba9, 7190fa4e3603c3b4ada132f9ee5bee4f2f92b2ef

### 6. InputHandler에서 errorHandler 변경
- 관련 커밋 : b3ad6f9090946c16504cec8d657d3fa0d5c0b7fa
- 기존 errorResults에 강하게 의존되어 있는 프로그램 구조를 변경했습니다! validator에서 바로 에러를 던지는 방식으로 변경하였고, InputHandler를 사용하는 대신 errorHandler 함수를 만들어 조금 더 단순하고 이해하기 쉽게 에러를 처리할 수 있도록 수정했습니다!

### 7. InputHandler에서 errorHandler 변경
- 관련 커밋 : b3ad6f9090946c16504cec8d657d3fa0d5c0b7fa, 1f7ca522c66530c9df747c69375013f2c46b7aa8
- 기존 errorResults에 강하게 의존되어 있는 프로그램 구조를 변경했습니다! validator에서 바로 에러를 던지는 방식으로 변경하였고, InputHandler를 사용하는 대신 errorHandler 함수를 만들어 조금 더 단순하고 이해하기 쉽게 에러를 처리할 수 있도록 수정했습니다!
- errorResults를 사용하지 않음으로써 프로그램의 전체 흐름이 많이 바뀌게 되었습니다.

### 8. 함수 집합의 객체는 카멜케이스로 변경
- 관련 커밋 : e9f70e31478af065b25631fe41d051da7b6b65e0
- 객체를 리턴하는 파일명을 카멜 케이스로 변경하였지만, 왜인지 모르겠는데 제대로 적용이 안되는 거 같아요 ㅜㅜ

### 9. 설계 수정
- Main 클래스는 전체 프로그램 흐름을 담당합니다. 이 클래스는 오직 LottoMachine과만 상호작용하며 비즈니스 문제를 해결합니다. 처음에는 Main 클래스에서 여러 클래스를 직접 사용하도록 구현했지만, 이로 인해 Main 클래스의 역할이 과중해지고 코드 이해도 어려워지는 문제가 있었습니다. 그래서 하나의 클래스에만 의존하도록 수정했습니다.
- LottoMachine은 Lottos와 Calculator 클래스를 활용하여 비즈니스 로직을 수행합니다. 각 클래스의 역할에 맞게 분리했는데, 혹시 더 분리하면 좋거나 부족한 부분이 있다면 말씀해 주세요!
- getter 사용 대신 메시지 전달 방식으로 필요한 데이터를 얻도록 구현하려고 노력했습니다. 이 부분 또한 보완할 점이 있다면 피드백 부탁드립니다!

---
# 🗺️설계
<img width="1581" alt="스크린샷 2025-02-23 00 47 30" src="https://github.com/user-attachments/assets/6b30c2b1-edb5-4831-83b0-abf45504acbb" />

LAH1203 commented on src/domain/validator/BonusNumberValidator.js:
```diff
@@ -0,0 +1,40 @@
+    this.#isDuplicated(winningNumbers, bonusNumber);
+  }
+
+  static #isNumberRangeOver(bonusNumber) {
+    if (
+      validationUtils.isNumberRangeOver(
+        bonusNumber,
+        LOTTO_NUMBER_RANGE.MIN,
+        LOTTO_NUMBER_RANGE.MAX
+      )
+    )
+      throw new Error(BONUS_NUMBER.IS_NUMBER_RANGE_OVER);
+  }
+
+  static #isNotNaturalNumber(bonusNumber) {
+    if (validationUtils.isNotNaturalNumber(bonusNumber))
+      throw new Error(BONUS_NUMBER.IS_NOT_NATURAL_NUMBER);
+  }
+
+  static #isDuplicated(winningNumbers, bonusNumber) {
```
private 메서드에 static을 붙이신 이유가 궁금합니다!
(나머지 validator도 비슷한 맥락)

yeongipark replied:
상태를 가지지 않는 Valiator 클래스의 메서드는 정적 메서드를 만들어서 인스턴스를 만들어서 사용하지 않아도 되도록 했습니다!

이에 퍼블릭 정적 메서드 내부에 프라이빗 메서드를 사용하러면 정적 메서드로 선언되어야 해서 프라이빗 메서드도 전부 정적 메서드로 선언했습니다..!

LAH1203 commented on src/domain/validator/BonusNumberValidator.js:
```diff
@@ -0,0 +1,40 @@
+};
+
+export default class BonusNumberValidator {
+  static bonusNumber(winningNumbers, bonusNumber) {
```
검증하는 메서드 이름은 bonusNumber보다 적절한 이름이 있을 것 같습니다~
(나머지 validator도 비슷한 맥락)

yeongipark replied:
처음에는 validateBonusNumber() 라는 메서드명을 사용했습니다.

이렇게 명명했을 때, BonusNumberValidator.validateBonusNumber() 로 사용되는데 이는 클래스 이름에서 보너스 번호 검증이라는 뜻이 포함되어 있는데 메서드에서도 validate를 사용하는 것은 의미가 중복된다고 생각하여 bonusNumber이라는 메서드 명을 사용하게 되었습니다.

예를 들어, output.printMessage()로 작성하는 것이 아니라 output.message() 처럼 말이죠

하지만 이런 논리면 bonusNumber라는 메서드명도 클래스에 이미 의미가 들어있으니 의미가 없네요.. 이 부분에 대해서는 다시 생각해봐야겠어요!

이 부분에 대해 하리는 어떻게 생각하는지 궁금해요!

yeongipark replied:
수정 커밋 : 9430ed672f8262ee7dfc87e1e85d602032c944f6

일단 수정했는데... 하리의 생각이 궁금해요!!

LAH1203 replied:
이전보다는 검증하는 함수라는 의미에서 훨씬 명확해진 것 같아요! `BonusNumberValidator.validateBonusNumber()`처럼 중복된 단어를 포함하여 호출하는 게 마음에 걸리신다면 그냥 `validate` 정도로만 해도 충분할 것 같아요 ㅎㅎ

LAH1203 commented on src/domain/validator/PurcahsePriceValidator.js:
```diff
@@ -0,0 +1,33 @@
+};
+
+const PURCHASE_PRICE = {
+  IS_NUMBER_RANGE_OVER: `[ERROR] ${PURCHASE_PRICE_RANGE.MIN} 이상 ${PURCHASE_PRICE_RANGE.MAX}억 이하의 숫자를 입력해야합니다.`,
```
10,000,000,000억 이하의 숫자인건가요...?!

yeongipark replied:
넵..!

LAH1203 replied:
허걱 1억만 받았으면...

LAH1203 commented on src/domain/validator/PurcahsePriceValidator.js:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  static #isNotMultiple(money) {
+    if (money % 1000 !== 0) throw new Error(PURCHASE_PRICE.IS_NOT_MULTIPLE);
```
이 1000 이란 단위도 상수이지 않을까요?

yeongipark replied:
1000을 상수로 뺐는데 놓친 부분이 있었군요..!

yeongipark replied:
수정 커밋 : 33ee81a2c207622c987653c783966723458ffa9a

LAH1203 commented on src/domain/validator/PurcahsePriceValidator.js:
파일명에 오타가 있습니다~!

yeongipark replied:
앗..! 수정하겠습니당

yeongipark replied:
수정 커밋 : c9db9d420aa5de71e5ce98c9f1bdbd9f97322cb7

LAH1203 commented on src/domain/Calculator.js:
```diff
@@ -0,0 +1,48 @@
+import { LOTTO_PRISE, RANK } from "../constant/definition.js";
+export default class Calculator {
+  // [{matchCount : 1, isMatchBonusNumber : true}]
+  static getStatistics(matchCounts) {
+    const rank = {
+      [RANK.FIFTH]: 0,
+      [RANK.FOURTH]: 0,
+      [RANK.THIRD]: 0,
+      [RANK.SECOND]: 0,
+      [RANK.FIRST]: 0,
+    };
+
+    matchCounts.forEach((matchData) => {
+      this.#judgeRank(matchData, rank);
+    });
+
+    return rank;
+  }
+
+  static #judgeRank({ matchCount, isMatchBonusNumber }, rank) {
+    switch (matchCount) {
+      case 3: {
+        rank[RANK.FIFTH]++;
+        break;
+      }
+      case 4: {
+        rank[RANK.FOURTH]++;
+        break;
+      }
+      case 5: {
+        if (isMatchBonusNumber) rank[RANK.SECOND]++;
+        else rank[RANK.THIRD]++;
+        break;
+      }
+      case 6: {
+        rank[RANK.FIRST]++;
+        break;
+      }
+    }
+  }
```
이런 식으로 배열의 특정 값에 1을 더하는 로직이 있고 거기에 더해 그 로직을 분리한 것은 로직의 전체적인 흐름을 파악하기에 방해가 되는 요소라고 생각합니다.
차라리 여기에도 `Array.reduce`를 활용하면 좀 더 흐름을 파악하기에 깔끔하도록 만들 수 있을 것 같은데, 어떠신가요?

yeongipark replied:
15줄을 넘지 않는 다는 요구 사항을 지키기 위해 분리했었습니다..!

reduce를 사용할 수 있다는 것을 까먹고 있었어요!!!!

reduce를 사용해서 깔끔하게 수정해보겠습니당

yeongipark replied:
수정 커밋 : 3619268c1d2a17d011f26343baf521ef8f1532f1

LAH1203 replied:
추가적으로, 지금은 switch를 사용해서 작성해주셨지만 저 경우에는 오히려 if 문이 뎁스를 줄일 수 있어서 가독성 측면에서 좋을 수 있어요 ㅎ.ㅎ 참고만 해주세요~~

LAH1203 commented on src/domain/Lottos.js:
Lottos와 LottoMachine이 다소 애매하게 분리된 것 같습니다. 서로 다른 객체로 분리를 해주셨다는 건 각각의 역할을 잘 수행해야 하는데, 두 객체는 서로 너무 강하게 결합되어 있기도 하고 Lottos의 역할 중 일부가 LottoMachine에 있는 경우도 있네요..

yeongipark replied:
수정 커밋 : 7025eb135766d2974e78c7e59fc32ffa54f47498

Lottos는 Lotto 인스턴스의 배열을 저장하는 클래스로 정답 번호와의 일치 여부 갯수, 로또들 번호 리턴을 수행합니다. 이를 클래스로 만들지 않고 LottoMachine에서 전부 수행하면 LottoMachine이 하는 역할이 너무 많다고 생각하여 Lottos라는 클래스를 만들어 분리했습니다!
> Lottos의 역할 중 일부가 LottoMachine에 있는 경우도 있네요..

은 Lotto 인스턴스를 만드는 작업이 Lottos에 있는게 맞다 생각하여 수정했는데, 애매하게 분리된 부분은 제가 보기에는 잘 모르겠어서,,, 수정하지 못했습니다..!

사실 아직 저만의 클래스를 분리하는 기준이 명확하게 없는게 가장 큰 문제인거 같습니다.. 저는 처음에 스파게티 코드로 작성하고 너무 많은 역할을 한다고 생각하면 그때서야 클래스를 분리하곤 합니다. 그리고 분리된 클래스의 필드의 값을 이용하는 로직은 해당 클래스에 작성하곤 합니다! 

혹시 하리만의 클래스를 분리하는 방법이 있으면 알려주세요 !!🙏

LAH1203 replied:
저는 처음에 설계에 시간을 많이 쏟는 편이에요!!!
일단 요구사항 명세대로 구현해야 할 기능을 나열해본 뒤에, 하나의 객체로 모을 수 있을 것 같은 기능끼리 묶어둡니당
그 중에서 객체가 여러 가지의 역할을 하는 것 같다면 또 한번 나눠보고요!
그렇게 처음 설계를 만든 다음에 구현하고, 구현했을 때 연관성이 크게 없는 로직이 포함되었다고 생각이 들면 그때마다 조금씩 분리하는 것 같아요

LAH1203 commented on src/util/validationUtil.js:
```diff
@@ -0,0 +1,6 @@
+const validationUtils = {
+  isNumberRangeOver: (number, min, max) => number < min || number > max,
```
min과 max는 range에 포함되지 않는걸까요?

yeongipark replied:
네!

해당 메서드를 호출하는 곳에서 매개변수로 min에 1, max에 45를 넘겨주고 있어서 range에 포함되지 않는 걸로 작성했습니다!

LAH1203 replied:
요건 제가 로직을 잘못 봤었네요...! 설명 감사합니다 🙌

LAH1203 commented on __tests__/BonusNumberValidator.test.js:
Validator에서 예외 케이스에 대한 테스트만 수행하고 있네요! 정상 케이스 (특히 경곗값) 에 대해서 테스트 케이스도 추가되면 좋을 것 같아요~

yeongipark replied:
수정 커밋 : 7234007828a52c4e376d5672fa7c45de1a537a54

LAH1203 commented on __tests__/Lotto.test.js:
```diff
@@ -0,0 +1,34 @@
+import Lotto from "../src/domain/Lotto.js";
+
+describe("Lotto 클래스 유효성 검사", () => {
+  test.each([
+    ["번호에 중복된 값이 있는 경우 예외가 발생한다.", [1, 2, 3, 4, 5, 5]],
+    ["번호의 갯수가 6개 아닌 경우 예외가 발생한다.", [1, 2, 3, 4, 5]],
+    [
+      "번호가 1~45 사이의 숫자가 아닌 경우 예외가 발생한다.",
+      [1, 2, 100, 4, 5, 6],
+    ],
+    ["번호가 자연수가 아닌 경우 예외가 발생한다.", ["a", 1, 2, 3, 4, 5]],
+  ])("%s", (_, numbers) => {
```
`test.each`를 가독성 있게 잘 적용해주셨네요 ! 짱짱

LAH1203 reviewed (changes requested):
우디 안녕하세요~ 구조가 많이 깔끔해졌네요! 👏
그래도 더 보완할 점이 있는 것 같아, 한번 더 수정 요청 드려봅니당 🥺

추가 질문 주신 부분부터 간단하게 답변 드려볼게요!

> 특정 기능을 구현하는 과정에서 여러 파일을 보다가, 갑자기 어떤 파일에 잘못된 코드가 있거나 리팩토링할 여지가 있는 코드를 발견할 때, 해당 파일만 수정해서 바로 커밋하는 편이 좋을까요? 아니면 수정할 부분은 나중에 미루고, 우선 현재 진행 중인 기능 구현을 마친 후에 별도로 수정하는 게 좋을까요??  
> 저는 매번 보일 때마다 수정해서 커밋이 뒤죽박죽이 되고 커밋 수도 많아지는 문제가 있는데, 이에 대해 어떻게 접근하는 게 좋은지 궁금합니다!

특정 기능을 구현하는 중이라면 구현 사항이 누락되지 않도록 진행 중인 구현을 마치고 별도로 수정하는 편이 좋을 것 같습니다~ 커밋 수가 많아지는 건 문제가 아니라고 생각해요! 다만 커밋 순서에 일관성이 없으면 보는 사람 입장에서도 이해가 어려울 수 있으니, 컨텍스트를 고려해서 남기시는 게 좋을 것 같아용

> 하리의 리뷰에서 재시작 부분이 도메인 로직이라고 하신 부분이 있었어요! 이 부분에 대해 저도 고민해 봤는데, 재시작 여부를 사용자에게 입력받고 처리하는 것은 도메인 로직에서 처리하기 보다는 애플리케이션 흐름을 담당하는 Main 클래스가 처리하는게 맞다고 생각하여, Main 클래스에 해당 로직을 작성했습니다!  
> 이 부분에 대해 하리의 생각이 궁금합니다!

이전 코멘트에서 우디는 `Main 클래스는 View와 Domain의 결합을 통해 프로그램 전체 흐름을 담당`한다고 말씀해주셨어요. 이 관점대로라면 Main은 단순히 View와 Domain을 결합만 해 프로그램의 흐름을 제어해야 할텐데, 왜 Main에 View와 Domain에서 가지고 있지 않은 고유의 로직이 포함되어야 할까요??
저는 이 부분이 의문이었습니다! 지금도 동일하고용

> 클래스에 메서드가 많아지면, 퍼블릭 메서드와 함께 여러 프라이빗 메서드도 등장하게 됩니다. 저는 가독성을 위해 퍼블릭 메서드에서 사용되는 프라이빗 메서드를 해당 퍼블릭 메서드 바로 아래에 배치하는 방식을 선호합니다. 하지만 전체 코드를 넓게 보았을 때, 퍼블릭과 프라이빗 메서드가 섞여 있으면 다소 난잡해 보일 수 있습니다. 이런 경우에는 프라이빗 메서드끼리 모으고, 퍼블릭 메서드끼리 모아서 배치하는 방법도 고려해볼 만하다고 생각합니다! 하리께서는 클래스 내부의 메서드 배치는 어떤 식으로 구성하시는지 궁금합니다!

저는 말씀해주신 두 규칙보다는 역할에 따라서 모아서 나열하는 걸 선호합니다.
예를 들면 input 관련 메서드, output 관련 메서드를 각각 모아서 나열하는 식이 될 수 있겠네요! 이 부분은 따로 베스트 프랙티스가 있는 게 아니어서, 나열 순서는 어떤 기준이든 일관성 있게 적용하면 되는 부분이라고 생각합니다~!

---

이외에 PR에서 남겨주신 일부 부분입니당

> 객체를 리턴하는 파일명을 카멜 케이스로 변경하였지만, 왜인지 모르겠는데 제대로 적용이 안되는 거 같아요 ㅜㅜ

깃에서는 대문자 <-> 소문자 변경을 제대로 인식하지 못해 이전 캐시가 그대로 올라가는 경우가 있어요~ 그래서 관련해서 설정을 해줘야 합니다!

#### 파일 대소문자 무시 설정 비활성화
```bash
git config core.ignorecase false
```

#### 캐시 제거
```bash
git rm -r --cached .
```

이후로 동일하게 add -> commit 하시면 됩니당

> 현재 유효성 검사는 여러 곳에서 수행되고 있습니다. Main 클래스에서는 사용자 입력에 대한 유효성 검사를 진행하지만, 특정 클래스 내부에서도 한 번 더 검증하는 이유는 예상치 못한 잘못된 입력으로부터 클래스 생성 자체를 막기 위함입니다. 프론트엔드는 보안에 취약한 영역이기 때문에, 클래스 내부에 유효성 검증 로직을 추가하여 보안을 강화했습니다.

말씀해주신 이 부분이 좀 더 납득이 가는 이유여야 할 것 같습니다ㅠㅠ
어차피 사용자 입력을 통해 유효성 검사를 통과한 값만 저장하는데, 그렇게 저장한 값에 대해서 한번 더 검증을 하는 이유가 잘못되었다고 생각합니다. 이미 저장한 값은 외부에서 접근할 수 없기 때문에 저장 당시에 검증이 되었다면 굳이 한번 더 동일한 검증을 할 필요는 없을 것 같아요.
또한 이미 생성된 클래스에서 한번 더 검증하기 때문에 그 클래스의 생성 자체를 막을 수도 없을 것 같아요!

---

그리고 제가 말씀 드렸던 커밋 번호를 남기는 건 각각의 피드백 코멘트에 reply로 다는 방식을 말씀 드렸던 거였습니다~! PR 코멘트로 따로 모아서 남기는 방식은 한 화면에서 이전 코멘트와 비교해서 보기가 힘들어서요 🥲
코드 리뷰 시 지켜지는 규칙(?) 같은 거여서 미리 익숙해지시면 좋을 것 같아 말씀드렸었는데, 이 부분은 나중에 추가 피드백 반영 시 참고해주세요 :)

yeongipark committed:
refactor(Calculator) : reduce를 이용해 로직 단순화

- getStatistics 메서드의 내부 로직에 reduce를 상용해 로직 단순화

yeongipark committed:
chore(PurchasePriceValidator) : 파일명 오타 수정

yeongipark committed:
refactor(PurchasePriceValidator) : 1000 상수화

yeongipark committed:
refactor: Lotto 배열 생성 로직 Lottos로 이동

yeongipark committed:
test: 정상 통과 테스트 로직 추가

yeongipark committed:
chore(view): 파일명 카멜 케이스로 변경

yeongipark committed:
refactor: validator 메서드 명 직관적으로 변경

yeongipark committed:
refactor(Main): 외부에서 사용되지 않는 메서드 프라이빗 메서드로 선언

yeongipark commented:
안녕하세요 @LAH1203  하리!🙂 
저의 부족한 코드를 꼼꼼하게 리뷰해 주셔서 감사합니다...! 덕분에 많은 고민을 하게 되었고, 평소에 아무 생각 없이 작성하던 부분도 생각하며 코드를 작성하게 되었어요!

하리의 리뷰를 토대로 다시 리팩토링을 진행했어요!!!

---

>말씀해주신 이 부분이 좀 더 납득이 가는 이유여야 할 것 같습니다ㅠㅠ
어차피 사용자 입력을 통해 유효성 검사를 통과한 값만 저장하는데, 그렇게 저장한 값에 대해서 한번 더 검증을 하는 이유가 잘못되었다고 생각합니다. 이미 저장한 값은 외부에서 접근할 수 없기 때문에 저장 당시에 검증이 되었다면 굳이 한번 더 동일한 검증을 할 필요는 없을 것 같아요.
또한 이미 생성된 클래스에서 한번 더 검증하기 때문에 그 클래스의 생성 자체를 막을 수도 없을 것 같아요!

하리 말씀처럼 제가 봐도 저의 논리가 부족하다고 생각이 들어요! 말을 덫 붙이자면, 유효성 검사를 통과한 값만 넘어와서 해당 애플리케이션에서는 잘못 생성될 경우가 없습니다. 따라서 클래스 내부에서 한번더 validate를 수행하는 것은 불필요하게 느껴질 수 있어요!

하지만, 객체의 불변성을 보장하고 무결성을 지키기 위해, 외부에서 이미 검증을 수행하더라도  클래스 자체에서 유효성 검사를 다시 한 번 수행하는 것이 좋다고 생각해요!

이에 대해 하리의 생각이 궁금합니당

--- 
> 그리고 제가 말씀 드렸던 커밋 번호를 남기는 건 각각의 피드백 코멘트에 reply로 다는 방식을 말씀 드렸던 거였습니다~! PR 코멘트로 따로 모아서 남기는 방식은 한 화면에서 이전 코멘트와 비교해서 보기가 힘들어서요 🥲
코드 리뷰 시 지켜지는 규칙(?) 같은 거여서 미리 익숙해지시면 좋을 것 같아 말씀드렸었는데, 이 부분은 나중에 추가 피드백 반영 시 참고해주세요 :)

이 부분에 대해서 이번에도 잘못했을까요.....? 😢

LAH1203 commented:
> 하리 말씀처럼 제가 봐도 저의 논리가 부족하다고 생각이 들어요! 말을 덫 붙이자면, 유효성 검사를 통과한 값만 넘어와서 해당 애플리케이션에서는 잘못 생성될 경우가 없습니다. 따라서 클래스 내부에서 한번더 validate를 수행하는 것은 불필요하게 느껴질 수 있어요!
> 하지만, 객체의 불변성을 보장하고 무결성을 지키기 위해, 외부에서 이미 검증을 수행하더라도 클래스 자체에서 유효성 검사를 다시 한 번 수행하는 것이 좋다고 생각해요!
> 이에 대해 하리의 생각이 궁금합니당

말씀해주신 의미에서는 괜찮은 설계라고 생각합니다~! 추가 설명 감사합니다 👍👍
저는 리액트의 부모 / 자식 컴포넌트 간의 관계에 대해 생각하면서 코멘트를 남겼던 건데, 요 부분은 지금 같이 인스턴스를 생성하는 측면에서는 우디가 말씀해주신 의견보다 설득력이 떨어지는 것 같네요...!

> 그리고 제가 말씀 드렸던 커밋 번호를 남기는 건 각각의 피드백 코멘트에 reply로 다는 방식을 말씀 드렸던 거였습니다~! PR 코멘트로 따로 모아서 남기는 방식은 한 화면에서 이전 코멘트와 비교해서 보기가 힘들어서요 🥲
> 코드 리뷰 시 지켜지는 규칙(?) 같은 거여서 미리 익숙해지시면 좋을 것 같아 말씀드렸었는데, 이 부분은 나중에 추가 피드백 반영 시 참고해주세요 :)
> 이 부분에 대해서 이번에도 잘못했을까요.....? 😢

헉... 울지 마세여..... 잘 남겨주셔서 감사합니다 🙇‍♂️🙇‍♂️ 덕분에 넘넘 편했어요
혹시 제가 너무 날카롭게 말하는 것 같은 부분이 있으면 슬쩍 말씀해주세요 ㅎㅎ

LAH1203 reviewed (approved):
1단계 너무너무 고생하셨습니다 우디~!
빠르게 피드백 반영해주시고 계속 더 나은 코드에 대해서 고민해주셔서, 덕분에 저도 점점 더 발전하는 코드를 보고 뿌듯했습니다 😆
2단계도 화이팅해서 부숴봅시당 💪
